### DISCUSSION TITLE
How to mock import.meta.env?

### DISCUSSION THREAD

**Discussion by koenwoortman** [vitest-dev/vitest#1606] [Q&A] [answered]
In my code I have a reference to `import.meta.env.MODE`. Can we mock this value in a test case?

Such that I could run a test as if the value of `import.meta.env.MODE` was X.

**Comment by sheremet-va** [comment, marked as answer]
```ts
import.meta.env.MODE = 'X'
```

**Reply by koenwoortman** [reply]
Tried this, but is doesn't work as I would expect for references in my code, only in the test files.

Lets for example say I have the following:
```js
// ./src/mode.js
export const mode = import.meta.env.MODE;
```
And to test the above use the following:
```js
// ./tests/mode.spec.js
import { it, expect } from 'vitest'
import { mode } from '../src/mode'

it('...', () => {
  import.meta.env.MODE = 'X'
  expect(mode).toBe('X')
});
```

I would expect this test to pass. However, `import.meta.env.MODE` from `./src/mode.js` returns `"test"` instead of `"X"`

```
 FAIL  tests/mode.spec.js > ...
AssertionError: expected 'test' to be 'X' // Object.is equality
 ❯ tests/mode.spec.js:6:15
      4| it('...', () => {
      5|        import.meta.env.MODE = 'X'
      6|        expect(mode).toBe('X')
       |               ^
      7| });
      8| 

  - Expected   "X"
  + Received   "test"
```

**Reply by sheremet-va** [reply]
Of course it would be, primitives are accessed by value, not by reference in JavaScript. You defined `mode` when `import.meta.env.MODE` was defined as `test`.

You need to import module _after_ you redefined it:
```ts
import.meta.env.MODE = 'x'
const { mode } = import('../src/mode')
expect(mode).toBe('X')
```

**Reply by sheremet-va** [reply]
If you for some reason have a module that exports all env like that, it would be easier to use `vi.mock` to mock full module:
```ts
import env from './envs'
vi.mock('./envs', () => ({ mode: 'X' }))
// you can also redefined it with
vi.spyOn(env, 'mode', 'get').mockReturnValue('NEW')
```

**Reply by koenwoortman** [reply]
@sheremet-va thanks!
